### ISSUE TITLE
MR approvers user/group has no effect (REST API change?)

### ISSUE THREAD

**Issue by weakcamel** [egnyte/gitlabform#68]
Applying a configuration using `approvers` and `approver_groups` has no effect while running against Gitlab 12.3.5-ee.

The config I'm using:

```
group_settings:
  'very-silly-test-group-for-ldap-and-fun':
    merge_requests:
      approvals:
        approvals_before_merge: 1
        reset_approvals_on_push: false
        disable_overriding_approvers_per_merge_request: false
        merge_requests_author_approval: true
      approver_groups:
        - gitlab
      approvers:
        - waldekm
        - foouser
```

With the log
```
$ gitlabform -vvv -c config.yml very-silly-test-group-for-ldap-and-fun
Reading config from: /Users/waldekm/git/config/config.yml
Connected to GitLab version: 12.3.5-ee (9dbaa740018)
Config file is declared to be compatible with GitLab API v4
Reading config from: /Users/waldekm/git/config/config.yml
*** # of groups to process: 1
*** # of projects to process: 1
* [1/1] Processing: very-silly-test-group-for-ldap-and-fun/testy-project
Setting project_settings
Setting project settings: {'only_allow_merge_if_pipeline_succeeds': True, 'only_allow_merge_if_all_discussions_are_resolved': True, 'shared_runners_enabled': True, 'request_access_enabled': True, 'wiki_enabled': False, 'snippets_enabled': False}
Setting merge_requests
Setting approvals settings: {'approvals_before_merge': 1, 'reset_approvals_on_push': False, 'disable_overriding_approvers_per_merge_request': False, 'merge_requests_author_approval': True}
Setting approvers to users ['waldekm', 'foouser'] and groups ['gitlab']
Setting hooks
```

Gitlabform implementation uses endpoint `projects/:id/approvers` which has been deprecated:
https://docs.gitlab.com/ee/api/merge_request_approvals.html#change-allowed-approvers

I could just be using it wrong (although I've tried moving the key `approvers` **under** `approvals` and this didn't work either). It might be also that the deprecated endpoint doesn't work anymore.  In any case, adding/removing users to the `approvers` list and running gitlabform doesn't show in `gitlab-rails/api_json.log`

**Comment by smcgregor83** [comment]
I am seeing the same behaviour running against 12.4.2-ee.

**Comment by weakcamel** [comment]
The new API - if powerful - is sadly much more complex and involves creating/updating *approval rules* (which have their own ids). 

https://docs.gitlab.com/ee/api/merge_request_approvals.html#create-project-level-rule

**Comment by weakcamel** [comment]
On a related note: https://gitlab.com/gitlab-org/gitlab/issues/4816 

TL;DR: with MR approvals enabled for a project, Gitlab assumes that all Developers should be notified on any MR. 

Gitlabform could be the most feasible workaround for this spammy behaviour.

**Comment by gdubicki** [comment]
Hi, thank you for reporting this and sorry for the trouble. I will try to get this fixed in the next 2-3 weeks.

**Comment by gdubicki** [comment]
I have tested this some time ago on 12.5.4-ee and have not notices this issue.

I have also tested this with the current  12.6.2-ee in https://travis-ci.com/egnyte/gitlabform/jobs/272436690 (done locally for now, will finish on Travis within a few minutes) and the problem does not exist there anymore.

Please see my test setup, configs and assertions in https://github.com/egnyte/gitlabform/pull/75/files#diff-c92a09cbae0c182205d0cbe7f414ff33 .

Can you please verify if this problem persists with GitLab >= 12.5.4, @weakcamel and @smcgregor83 ?

**Comment by gdubicki** [comment]
Ok, so my latests test confirm that there is an issue if you have both users AND groups set as approvers and change any of these lists, for both `12.5.4` and `12.6.2`.

(There is NO such problem if you ONLY have users OR only groups as approvers and change them.)

We need to report this to GitLab as even if this API is deprecated it should not have such bugs.

**Comment by weakcamel** [comment]
I've noticed a very inconsistent behaviour here:

* if there is *no* validation rule, gitlabform run doesn't apply anything, no matter what
* if I add a validation rule manually and it contains a group only - my update has no effect, which confirms your observations @gdubicki 
* occasionally by re-running gitlabform on the project and manually tweaking the same validation rule - repeated a few times - seems to unblock it sometimes, although I've not managed to make it repeatable yet

So yes, I suspect that Gitlab implemented something which used to be consistent with their previous design (or most common setup) for the deprecated API...

**Comment by gdubicki** [comment]
Sorry for the delay.

I started to work on implementing the new API but I have very strange results for now.

I suspect that I may be hitting https://gitlab.com/gitlab-org/gitlab/issues/36014 , https://gitlab.com/gitlab-org/gitlab/issues/195793 or some similar GitLab bug.

**Comment by gdubicki** [comment]
According to my tests the latest GitLab only sets the first user and the first group from the lists provided in https://docs.gitlab.com/ee/api/merge_request_approvals.html#update-project-level-rule requests .

However I will check how it is done in the GitLab provider for Terraform, which almost has this feature now too (https://github.com/terraform-providers/terraform-provider-gitlab/pull/208/files). Maybe I am doing something wrong?

**Comment by gdubicki** [comment]
I think that the provider's PR tests do not cover the case where I detected the problem. I reported that in that PR already.

I also reported the bug to GitLab, here: https://gitlab.com/gitlab-org/gitlab/issues/198770

**Comment by gdubicki** [comment]
I have read GitLab code and if I got it right then the tests for the API there do NOT cover case with more than 1 user or more than 1 group set as approvers: https://gitlab.com/gitlab-org/gitlab/blob/master/ee/spec/requests/api/project_approval_rules_spec.rb

Setting more users or groups in the GitLab web UI DOES work but it is using a different code than that API - there is a ticket open to switch it to use the public API here: https://gitlab.com/gitlab-org/gitlab/issues/13574

But I find it hard to believe that it could be possible that this bug in fact exists and no one noticed it for such a long time since the API is available, so I think that I rather have a bug in my code. 😞

**Comment by gdubicki** [comment]
Unfortunately python-gitlab does not have this implemented so I can’t check how they have dealt with it: https://github.com/python-gitlab/python-gitlab/issues/877

**Comment by gdubicki** [comment]
I have come up with a temporary **workaround** for this bug:

1. Create GitLab group such as "approvers_group_a" and in GitLabForm assign them members.
2. Assign ONLY THIS GROUP to your appropriate projects as approver groups with GitLabForm.

Example:

```
# (...)
group_settings:
  approver_group_a:
    members:
      - user1
      - user2

project_settings:
  some_group/project_a:
    merge_requests:
      approvals:
        approvals_before_merge: 1
        reset_approvals_on_push: false
        disable_overriding_approvers_per_merge_request: true
      approver_groups:
        - approver_group_a
```

I know that this is not pretty but it should work until this bug is resolved.

**Comment by gdubicki** [comment]
Version 1.11.0 is just being built and pushed to PyPI. I contains the group members adding/changing access level (I am trying to be precise here as we cannot yet remove members with gitlabform). So the above workaround should work now. :)

**Comment by gdubicki** [comment]
Version 1.12.0 is just being built and pushed to PyPI. It contains new Merge Request Approvals API implementation, so the code above should work now.

**Comment by weakcamel** [comment]
Many many thanks @gdubicki 

I can confirm that the latest version works within limitations caused by https://gitlab.com/gitlab-org/gitlab/issues/198770 (when I tried, I got the lovely warning about it.

I've voted for the [issue](https://gitlab.com/gitlab-org/gitlab/issues/198770) and now commented on it as well, fingers crossed it gets some interest from Gitlab people soon.

**Comment by gdubicki** [comment]
I am just releasing 1.13.0 with full group membership management, including removing users from a group. But note that with that you HAVE TO have at least one Owner in each group, GitLab requires that! Please see updated syntax in the comment above.

**Comment by gdubicki** [comment]
@weakcamel and everyone affected: I am happy to report that the issue is resolved in v1.14.0 that has been released just now.

(But less-than-happy to learn that it was caused by something that I consider a GitLab issue more than a GitLabForm issue - see #90 and the linked GitLab issue for more info.)

**Comment by weakcamel** [comment]
@gdubicki  this is amazing! I'm definitely going to give it a try as soon as I can :)

**Comment by weakcamel** [comment]
Works like a charm!

👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍
